rest-server/core/src/main/scala/com/actionml/core/spark/SparkContextSupport.scala

```diff
@@ -17,49 +17,74 @@
 
 package com.actionml.core.spark
 
-import cats.data.Validated
-import cats.data.Validated.{Invalid, Valid}
-import com.actionml.core.store.backends.MongoStorage
-import com.actionml.core.validate._
+import java.util.concurrent.atomic.AtomicReference
+
 import com.typesafe.scalalogging.LazyLogging
 import org.apache.spark.{SparkConf, SparkContext}
 import org.json4s._
 import org.json4s.jackson.JsonMethods._
 
+import scala.concurrent.ExecutionContext.Implicits.global
+import scala.concurrent.{Future, Promise}
+import scala.reflect.ClassTag
+import scala.util.Try
+
 
-trait SparkContextSupport extends LazyLogging {
-  import com.actionml.core.spark.SparkContextSupport._
+trait SparkContextSupport[A] {
+  this: SparkStorageSupport[A] with LazyLogging =>
+  import SparkContextSupport._
+
+  def execute[B](fn: Iterator[A] => B, config: String, defaults: Map[String, String])(implicit ca: ClassTag[A], cb: ClassTag[B]): Future[Array[B]] = {
```

Comment:
Ideally this should execute a block of code in the scope of the calling function. To the Scala compiler this would be an anonymous function whose input is whatever variables in scope that it references.  Can we do something like:

```
sparkContext.execute { 
  ...
}
where the block is executed in a new thread?

Reply:
we need an id for context. For now, it's config + defaults. I can change execute signature to
``` def execute[B](config: String, defaults: Map[String, String])(fn: Iterator[A] => B)(implicit ca: ClassTag[A], cb: ClassTag[B]): Future[Array[B]]``` and call will became 
```execute(initParams, defaults) { case i: Iterator[Document] =>
        ???
    }``` is it ok?

Reply:
resolved